File: notifications-polyglot/golang/adapter/adapter_test.go
package adapter

import (
	"testing"
)

func TestNewChannelFactory(t *testing.T) {
	factory := NewChannelFactory()
	if factory == nil {
		t.Fatal("Factory should not be nil")
	}
	if len(factory.adapters) != 3 {
		t.Fatalf("Expected 3 adapters, got %d", len(factory.adapters))
	}
}

func TestGetAdapter_Email(t *testing.T) {
	factory := NewChannelFactory()
	adapter, err := factory.GetAdapter("email")
	if err != nil {
		t.Fatalf("Failed to get email adapter: %v", err)
	}
	if !adapter.Supports("email") {
		t.Error("Email adapter should support 'email' channel")
	}
}

func TestGetAdapter_SMS(t *testing.T) {
	factory := NewChannelFactory()
	adapter, err := factory.GetAdapter("sms")
	if err != nil {
		t.Fatalf("Failed to get sms adapter: %v", err)
	}
	if !adapter.Supports("sms") {
		t.Error("SMS adapter should support 'sms' channel")
	}
}

func TestGetAdapter_Push(t *testing.T) {
	factory := NewChannelFactory()
	adapter, err := factory.GetAdapter("push")
	if err != nil {
		t.Fatalf("Failed to get push adapter: %v", err)
	}
	if !adapter.Supports("push") {
		t.Error("Push adapter should support 'push' channel")
	}
}

func TestGetAdapter_Unsupported(t *testing.T) {
	factory := NewChannelFactory()
	_, err := factory.GetAdapter("voice")
	if err == nil {
		t.Error("Expected error for unsupported channel")
	}
}

func TestEmailAdapter_Send(t *testing.T) {
	adapter := &EmailAdapter{}
	err := adapter.Send("user@example.com", "Test message")
	if err != nil {
		t.Errorf("Send should not return error: %v", err)
	}
}

func TestSmsAdapter_Send(t *testing.T) {
	adapter := &SmsAdapter{}
	err := adapter.Send("+1234567890", "Test message")
	if err != nil {
		t.Errorf("Send should not return error: %v", err)
	}
}

func TestPushAdapter_Send(t *testing.T) {
	adapter := &PushAdapter{}
	err := adapter.Send("device-token-123", "Test message")
	if err != nil {
		t.Errorf("Send should not return error: %v", err)
	}
}

File: notifications-polyglot/rust/tests/adapter_test.rs
use notification_service_rust::adapter::{ChannelFactory, EmailAdapter, SmsAdapter, PushAdapter, ChannelAdapter};

#[test]
fn test_channel_factory_new() {
    let factory = ChannelFactory::new();
    // Factory should have 3 adapters registered
    assert_eq!(factory.get_adapter("email").is_some(), true);
    assert_eq!(factory.get_adapter("sms").is_some(), true);
    assert_eq!(factory.get_adapter("push").is_some(), true);
}

#[test]
fn test_get_adapter_email() {
    let factory = ChannelFactory::new();
    let adapter = factory.get_adapter("email");
    assert!(adapter.is_some(), "Should return email adapter");
}

#[test]
fn test_get_adapter_sms() {
    let factory = ChannelFactory::new();
    let adapter = factory.get_adapter("sms");
    assert!(adapter.is_some(), "Should return sms adapter");
}

#[test]
fn test_get_adapter_push() {
    let factory = ChannelFactory::new();
    let adapter = factory.get_adapter("push");
    assert!(adapter.is_some(), "Should return push adapter");
}

#[test]
fn test_get_adapter_unsupported() {
    let factory = ChannelFactory::new();
    let adapter = factory.get_adapter("voice");
    assert!(adapter.is_none(), "Should return None for unsupported channel");
}

#[tokio::test]
async fn test_email_adapter_send() {
    let adapter = EmailAdapter;
    let result = adapter.send("user@example.com", "Test message").await;
    assert!(result.is_ok(), "Send should succeed");
}

#[tokio::test]
async fn test_sms_adapter_send() {
    let adapter = SmsAdapter;
    let result = adapter.send("+1234567890", "Test message").await;
    assert!(result.is_ok(), "Send should succeed");
}

#[tokio::test]
async fn test_push_adapter_send() {
    let adapter = PushAdapter;
    let result = adapter.send("device-token-123", "Test message").await;
    assert!(result.is_ok(), "Send should succeed");
}
